service/grpc/server.go

```diff
@@ -18,53 +18,69 @@ package grpc
 
 import (
 	"context"
-	"sync"
 
+	"github.com/google/uuid"
+	"github.com/jellydator/ttlcache/v3"
 	"github.com/plgd-dev/client-application/pb"
 	serviceDevice "github.com/plgd-dev/client-application/service/device"
+	"github.com/plgd-dev/client-application/service/remoteProvisioning"
+	"github.com/plgd-dev/go-coap/v2/pkg/sync"
 	"github.com/plgd-dev/hub/v2/pkg/log"
+	"go.uber.org/atomic"
 	"google.golang.org/grpc/codes"
 	"google.golang.org/grpc/status"
 )
 
 type ClientApplicationServer struct {
-	serviceDevice *serviceDevice.Service
-	info          *ServiceInformation
-	logger        log.Logger
-	devices       sync.Map
 	pb.UnimplementedClientApplicationServer
+
+	serviceDevice            *serviceDevice.Service
+	info                     *ServiceInformation
+	logger                   log.Logger
+	devices                  *sync.Map[uuid.UUID, *device]
+	csrCache                 *ttlcache.Cache[uuid.UUID, bool]
+	remoteProvisioningConfig remoteProvisioning.Config
+	jwksCache                atomic.Pointer[JSONWebKeyCache]
+	remoteOwnSignCache       *sync.Map[uuid.UUID, *remoteSign]
 }
 
-func NewClientApplicationServer(serviceDevice *serviceDevice.Service, info *ServiceInformation, logger log.Logger) *ClientApplicationServer {
+func NewClientApplicationServer(remoteProvisioningConfig remoteProvisioning.Config, serviceDevice *serviceDevice.Service, info *ServiceInformation, logger log.Logger) *ClientApplicationServer {
+	csrCache := ttlcache.New[uuid.UUID, bool]()
+	remoteOwnSignCache := ttlcache.New[uuid.UUID, *remoteSign]()
+	go csrCache.Start()
+	go remoteOwnSignCache.Start()
 	return &ClientApplicationServer{
-		serviceDevice: serviceDevice,
-		info:          info,
-		logger:        logger,
+		serviceDevice:            serviceDevice,
+		info:                     info,
+		logger:                   logger,
+		csrCache:                 csrCache,
+		remoteProvisioningConfig: remoteProvisioningConfig,
+		remoteOwnSignCache:       sync.NewMap[uuid.UUID, *remoteSign](),
+		devices:                  sync.NewMap[uuid.UUID, *device](),
 	}
 }
 
 func (s *ClientApplicationServer) Version() string {
 	return s.info.Version
 }
 
-func (s *ClientApplicationServer) getDevice(deviceID string) (*device, error) {
-	d, ok := s.devices.Load(deviceID)
+func (s *ClientApplicationServer) Close() {
+	s.csrCache.Stop()
```

Comment:
Doesn't s.remoteOwnSignCache need to be stopped as well?

Reply:
no it is `sync.Map` type